app/javascript/lib/metrics-perf.js

```diff
@@ -171,19 +181,63 @@ export function calculatePerfTimes(perfTimes) {
       // so, e.g. the cluster plot will wait until after the expression data is fetched, and then render itself
       // based on a subset of that data
       'perfTime:frontend:other': now - perfTimes.requestStart - perfTimes.plot,
-      'perfTime:frontend:plot': perfTimes.plot,
-      'cache': 'client'
+      'perfTime:frontend:plot': perfTimes.plot
     }
     const perfProps = roundValues(Object.assign({}, rawPerfProps))
     perfProps['perfTime:url'] = perfTimes.url
+
+    perfProps['perfTime:cache'] = 'client-plot-data'
+    return perfProps
+  } else if (perfTimes.serviceWorkerCacheHit) {
+    // Processing SCP's generic frontend service worker caching
+
+    let rawPerfProps = {
+      'perfTime:full': now - perfTimes.requestStart,
+      'perfTime': now - perfTimes.requestStart,
+      'perfTime:legacy': now - perfTimes.requestStart,
+
+      // Server timing
+      'perfTime:backend': STEP_NOT_NEEDED,
+
+      // Client timing
+      'perfTime:frontend': now - perfTimes.requestStart, // Time from API call response start to effect (e.g. plot) end
+      'perfTime:frontend:transfer': STEP_NOT_NEEDED,
+      'perfTime:frontend:parse': perfTimes.parse
+    }
+
+    if ('plot' in perfTimes) {
+      rawPerfProps = Object.assign({
+      // note that 'other' here is very likely to include the rendering of a different plot.
+      // if a plot is fully cached, that usually means two plots were requested, one of which is a subset of the other
+      // so, e.g. the cluster plot will wait until after the expression data is fetched, and then render itself
+      // based on a subset of that data
+        'perfTime:frontend:other': now - perfTimes.requestStart - perfTimes.plot,
+        'perfTime:frontend:plot': perfTimes.plot
+      }, rawPerfProps)
+    }
+
+    let perfProps = roundValues(Object.assign({}, rawPerfProps))
+
+    perfProps = Object.assign({
+      'perfTime:cache': 'service-worker',
+      'perfTime:serviceWorkerCacheHit': true,
+      'perfTime:url': perfTimes.url
+    }, perfProps)
     return perfProps
   }
 
   const plot = perfTimes.plot ? perfTimes.plot : 0
 
-  const perfEntry =
-    performance.getEntriesByType('resource')
-      .filter(entry => entry.name === perfTimes.url)[0]
+  let perfEntry = null
+  const resourcePerfEntries = performance.getEntriesByType('resource')
```

Comment:
Good question!  It caught my eye because `filter` is clearly suboptimal, then I assumed `for` would be faster than `find` based on `for` often being faster than newer equivalents.  It turns out [`.find()` is faster than `for...break`](https://stackoverflow.com/a/50843994), so I'll use that.